nimbus/src/enrollment.rs

```diff
@@ -20,28 +23,300 @@ pub enum EnrolledReason {
+                            &branch,
+                            Some("targeting"),
+                            EnrollmentChangeEventType::Disqualification,
+                        ));
+                        Self {
+                            slug: self.slug.clone(),
+                            status: EnrollmentStatus::Disqualified {
+                                reason: DisqualifiedReason::Targeting,
+                                enrollment_id,
+                                branch,
+                            },
+                        }
+                    }
+                }
+            }
+            EnrollmentStatus::Disqualified { .. } => {
+                if !is_user_participating {
+                    log::debug!(
+                        "Disqualified experiment enrollment '{}' has been discarded (global opt-out)",
+                        &self.slug
+                    );
+                    // Discard disqualified enrollments on global experiments opt-out.
+                    return Ok(None);
+                }
+                self.clone()
+            }
+            EnrollmentStatus::WasEnrolled { .. } => self.clone(),
+            EnrollmentStatus::Error { .. } => self.clone(), // Cannot recover from errors!
+        }))
+    }
+
+    fn on_experiment_discarded(
+        &self,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Option<Self> {
+        log::debug!(
+            "Experiment '{}' vanished while we had enrollment status of {:?}",
+            self.slug,
+            self
+        );
+        let (branch, enrollment_id) = match &self.status {
+            EnrollmentStatus::Enrolled {
+                branch,
+                enrollment_id,
+                ..
+            } => (branch, enrollment_id),
+            EnrollmentStatus::Disqualified {
+                branch,
+                enrollment_id,
+                ..
+            } => (branch, enrollment_id),
+            _ => return None, // We were never enrolled, bail.
+        };
+        enrollment_events.push(EnrollmentChangeEvent::new(
+            &self.slug,
+            enrollment_id,
+            branch,
+            None,
+            EnrollmentChangeEventType::Unenrollment,
+        ));
+        Some(Self {
+            slug: self.slug.clone(),
+            status: EnrollmentStatus::WasEnrolled {
+                enrollment_id: enrollment_id.to_owned(),
+                branch: branch.to_owned(),
+                observed_at: {
+                    SystemTime::now()
+                        .duration_since(UNIX_EPOCH)
+                        .expect("Current date before Unix Epoch.")
+                        .as_secs()
+                },
+            },
+        })
+    }
+}
+
 #[derive(Deserialize, Serialize, Debug, Clone, Hash, Eq, PartialEq)]
 pub enum EnrollmentStatus {
     // Enrolled
     Enrolled {
+        enrollment_id: String, // Random ID used for telemetry events correlation.
```

Comment:
why not a `Uuid` (here and below)? Is it due to uniffi? If so, we have that problem in forms in a-s - so maybe a comment here, so it's obvious it should change should uniffi get better support?